pkg/test/envoy/envoy.go

```diff
@@ -111,9 +114,65 @@ func (e *Envoy) Start() (err error) {
 	// Run the envoy binary
 	args := e.getCommandArgs()
 	e.cmd = exec.Command(envoyPath, args...)
-	e.cmd.Stderr = os.Stderr
+
+	// Start the log interceptor for Envoy stderr
+	startMutex := sync.Mutex{}
+	starting := true
+	logCh := make(chan string)
+	errorCh := make(chan error)
+	startComplete := func() {
+		startMutex.Lock()
+		defer startMutex.Unlock()
+		starting = false
+		close(logCh)
+		close(errorCh)
+	}
+	defer startComplete()
+	runIfStarting := func(fn func()) {
+		startMutex.Lock()
+		defer startMutex.Unlock()
+		if starting {
+			fn()
+		}
+	}
+	e.stderrInterceptor, err = newLogInterceptor(os.Stderr,
+		func(logLine string) {
+			runIfStarting(func() {
+				logCh <- logLine
+			})
+		},
+		func(err error) {
+			runIfStarting(func() {
+				errorCh <- err
+			})
+		})
+	if err != nil {
+		return err
+	}
+	e.stderrInterceptor.start()
+
+	// Redirect stderr from Envoy to the interceptor
+	e.cmd.Stderr = e.stderrInterceptor.Writer()
 	e.cmd.Stdout = os.Stdout
-	return e.cmd.Start()
+	if err := e.cmd.Start(); err != nil {
+		return err
+	}
+
+	// Wait for a period, checking the Envoy logs for any startup errors.
+	timer := time.NewTimer(1 * time.Second)
+	for {
+		select {
+		case <-timer.C:
+			return nil
+		case logLine := <-logCh:
+			// For now, address binding is the main issue. We can expand this in the future as needed.
+			if strings.Contains(logLine, "Address already in use") {
```

Comment:
1. Wouldn't envoy just exit 1 if a bootstrap address is in use?
2. Is the goal here to wait for a ready envoy? If so, then detecting an error is not sufficient, since envoy can take awhile initializing its data structures (clusters, etc) in a loaded system.

Reply:
E.g. see https://mail.google.com/mail/u/0/#all/FMfcgxwBVWQNkBbftgXpnHdbdTVWglSD

Reply:
E.g. see https://github.com/envoyproxy/envoy/issues/5862

Reply:
> 1. Wouldn't envoy just exit 1 if a bootstrap address is in use?

No, it doesn't seem to.

> 2. Is the goal here to wait for a ready envoy? If so, then detecting an error is not sufficient, since envoy can take awhile initializing its data structures (clusters, etc) in a loaded system.

That's fair. The attempt here was to account for the fact that occasionally `listener.Close()` is not enough to unbind a port before handing it over to Envoy. My initial thought was that the test could simply retry the process of port selection and starting Envoy if we can detect binding errors on startup. However, we also bind new ports as listeners are added dynamically, so this still wouldn't account for all the cases.

Let me close this ... not really the right solution.